## Projet final: Cartographie des Territoires à Potentiel de Flexibilité Énergétique.

### Introduction: 

Dans un contexte de transition énergétique et de tension croissante sur les réseaux électriques français, les entreprises telles que EDF, Engie, RTE ou TotalEnergies doivent développer des stratégies permettant de moduler la demande d’électricité. La flexibilité énergétique (c’est-à-dire la capacité à décaler ou réduire temporairement la consommation ) devient un levier essentiel pour :

- Soulager les réseaux en période de pointe,
- Limiter les risques de délestage,
- Réduire les coûts d’équilibrage du système,
- Et mieux valoriser les contrats d’effacement.

Dans ce contexte, l’entreprise FlexiMap Solutions a été choisie pour réaliser une étude approfondie sur les territoires ayant un fort potentiel en matière de flexibilité énergétique. Concrètement, il s’agit d’identifier les zones en France où il serait le plus pertinent de mettre en place des dispositifs permettant de moduler ou de réduire temporairement la consommation d’énergie, par exemple lors des pics de demande. Pour mener ce projet à bien, FlexiMap travaille en étroite collaboration avec les gestionnaires de réseaux (comme RTE ou Enedis) et les collectivités locales.

#### Qui est FlexiMap Solutions ?

FlexiMap Solutions est une jeune entreprise innovante qui se spécialise dans l’analyse de données énergétiques à l’échelle locale et dans la création de cartes et d’outils visuels destinés à faciliter la prise de décision. Son objectif est d’aider les acteurs de la transition énergétique – qu’ils soient publics (collectivités, agences) ou privés (entreprises, opérateurs d’énergie) – à mieux comprendre leur territoire et à cibler leurs actions de manière efficace.

Concrètement, FlexiMap intervient dans plusieurs domaines :

- Elle aide à repérer les zones où la consommation d’énergie peut être modulée facilement, en particulier dans les secteurs industriel et tertiaire.

- Elle identifie les opportunités pour réduire la consommation temporairement (ce qu’on appelle l’effacement), que ce soit chez les particuliers ou dans les entreprises.

- Elle propose des analyses qui permettent de décider où investir en priorité pour renforcer les réseaux électriques ou mettre en place des solutions de pilotage.

- Enfin, elle conçoit des outils pratiques comme des cartes interactives ou des tableaux de bord, qui rendent les données lisibles et exploitables par les décideurs.

Grâce à une approche rigoureuse et fondée sur l’analyse de données concrètes, FlexiMap fournit des résultats clairs, directement utilisables par les acteurs du secteur pour orienter leurs politiques énergétiques ou leurs investissements.


#### Problématique:

##### Quels territoires présentent une forte consommation dans les secteurs “pilotables” (tertiaire et industriel) et pourraient ainsi être ciblés en priorité pour des programmes de flexibilité ou d’effacement énergétique ?

#### Objectifs de l’analyse:

##### 1- Identifier les EPCI et départements :

- avec une consommation significative dans les secteurs tertiaire et industriel,
- avec un déséquilibre ou une dépendance à une source (élec/gaz) qui pourrait renforcer le besoin de modulation.

##### 2- Construire un indice de flexibilité basé sur la part pilotable dans la consommation totale.

##### 3- Créer une cartographie des zones à cibler en priorité pour les contrats de flexibilité ou les solutions de pilotage énergétique.

Jeux de données utilisés:

##### - conso-departement-annuelle.csv : consommation d’énergie par département
##### - conso-epci-annuelle.csv : consommation d’énergie par intercommunalité (EPCI)

Ces jeux de données sont consultables en open source sur les sites officiels suivants:

##### https://www.data.gouv.fr/fr/ 

##### https://opendata.reseaux-energies.fr/


## Première partie: données et préparation

#### Objectif: automatiser le nettoyage, la structuration et le calcul des indicateurs clés.

#### Importation des bibliothèques

In [20]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path
import csv
import os
from collections import Counter

#### 1.1 Présentation du jeu de données

#### Commentaire: 

##### ❌ Problème de chargement des fichiers CSV

Lors de l’ouverture des fichiers conso-departement-annuelle.csv et conso-epci-annuelle.csv, une erreur de lecture s’est produite sur le premier fichier, indiquant qu'il y'avait un problème d'encoding et de séparateur.


#### Solution: 

Les fichiers utilisent le point-virgule (;) comme séparateur, ce qui est courant en France, alors que la fonction de lecture intégrée dans pandas ne le traite pas automatiquement.

#### Chargement des données

In [26]:
# Chargement des fichiers avec le bon séparateur
df_departement = pd.read_csv(r"conso-departement-annuelle.csv", sep=';',encoding="utf-8")
df_epci = pd.read_csv(r"conso-epci-annuelle.csv", sep=';',encoding="utf-8")

# Affichage des premières lignes pour vérification
df_epci.head()


Année  Code EPCI                   Libellé EPCI Code département  \
0   2020  245200597            CC des Trois Forêts               52   
1   2020  245300389         CC du Bocage Mayennais               53   
2   2020  245300447  CC du Pays de Château-Gontier               53   
3   2020  245400601         CC du Bassin de Pompey               54   
4   2020  245500327              CC du Sammiellois               55   

  Libellé département Code région    Libellé région  \
0         Haute-Marne          44         Grand Est   
1             Mayenne          52  Pays de la Loire   
2             Mayenne          52  Pays de la Loire   
3  Meurthe-et-Moselle          44         Grand Est   
4               Meuse          44         Grand Est   

   Consommation agriculture (MWh)  Consommation industrie (MWh)  \
0                     1625.588036                  11350.755349   
1                    21235.242249                  46520.384339   
2                    14227.505347                 112967.295131   
3                     3086.327198                 517458.677701   
4                    56176.485376                  93984.196303   

   Consommation résidentiel (MWh)  ...  Opérateur électricité  \
0                    22146.194547  ...                 Enedis   
1                    55405.231422  ...             RTE,Enedis   
2                   125315.057404  ...                 Enedis   
3                   251200.004311  ...             RTE,Enedis   
4                    40639.475396  ...                 Enedis   

   Consommation gaz agriculture (MWh)  Consommation gaz industrie (MWh)  \
0                             0.00000                           0.00000   
1                             0.00000                           0.00000   
2                           782.23752                       54661.60757   
3                            54.47792                      326828.51579   
4                         53983.81400                       66962.37418   

  Consommation gaz résidentiel (MWh)  Consommation gaz tertiaire (MWh)  \
0                            0.00000                           0.00000   
1                            0.00000                           0.00000   
2                        45984.23471                       37956.19229   
3                       162693.49501                       36460.67458   
4                        19807.20546                       10925.37091   

   Consommation gaz autre (MWh)  Consommation gaz totale (MWh)  Opérateur gaz  \
0                           0.0                        0.00000        (Aucun)   
1                           0.0                        0.00000        (Aucun)   
2                           0.0                   139384.27209    GRDF,NaTran   
3                           0.0                   526037.16330    GRDF,NaTran   
4                           0.0                   151678.76455           GRDF   

                                      Géo-shape EPCI  \
0  {"coordinates": [[[[4.954099565, 47.866767159]...   
1  {"coordinates": [[[[-0.780693738, 48.295050524...   
2  {"coordinates": [[[[-0.434123887, 47.809234237...   
3  {"coordinates": [[[[6.20119471, 48.729226945],...   
4  {"coordinates": [[[[5.414532493, 48.799938714]...   

               Géo-point EPCI  
0   48.020292005, 4.938389581  
1  48.416569068, -0.824073812  
2  47.822865292, -0.668455364  
3   48.767444901, 6.122330505  
4   48.916066813, 5.514500311  

[5 rows x 30 columns]

#### Analyse: 

##### Tableau 1 : Consommation départementale annuelle

Ce tableau contient de nombreuses colonnes (28), parfois techniques ou redondantes. Les plus utiles sont :
- Identifiants : Année, Code et Libellé du département, région
- Consommation par secteur (électricité + gaz) : agriculture, industrie, résidentiel, tertiaire, autre
- Opérateurs : électricité et gaz

##### Problèmes relevés :

- Données absentes ou incohérentes : plusieurs départements affichent une consommation de gaz à 0, ce qui semble peu crédible. Cela peut venir d’erreurs, d’un réseau de gaz inexistant ou d’un oubli de saisie.
- Contradictions sur les opérateurs : parfois, un opérateur est noté "Aucun" alors que des consommations de gaz sont indiquées.
- Noms d’opérateurs multiples : plusieurs opérateurs listés ensemble compliquent l’analyse automatique.
- Colonnes géographiques lourdes : utiles pour la cartographie, mais encombrantes si l’on analyse seulement les chiffres.

##### Tableau 2 : Consommation annuelle par EPCI

Structure similaire au tableau précédent, avec beaucoup de détails. Colonnes principales :
- Identifiants : Année, Code et Libellé de l’EPCI, département, région
- Consommation par secteur (électricité + gaz)
- Opérateurs et coordonnées géographiques

##### Problèmes relevés :

- Consommation de gaz à zéro, mais opérateur présent : cela révèle probablement des erreurs ou des données non mises à jour.
- Tableau dense : riche en données, mais nécessite de filtrer pour rester lisible.
- Forte variabilité entre territoires : certains EPCI consomment surtout dans l’industrie, d’autres dans le résidentiel. Les comparaisons sans contexte peuvent être trompeuses.
- Format géographique complexe : certaines colonnes sont difficiles à exploiter sans outils de cartographie.
- Noms d’EPCI hétérogènes : noms parfois longs ou mal normalisés, ce qui peut compliquer les recherches ou tris.

#### III Nettoyage préliminaire des données

#### A- Vérification structurelle des en-tetes

In [31]:
# Fonction pour détecter les problèmes structurels dans les en-têtes
def check_header_issues(df, df_name):
    issues = {}

    # Espaces en début ou fin de nom de colonne
    cols_with_spaces = [col for col in df.columns if col != col.strip()]
    if cols_with_spaces:
        issues['espaces_dans_noms_colonnes'] = cols_with_spaces

    # Colonnes sans nom (vide ou nom générique)
    unnamed_cols = [col for col in df.columns if 'unnamed' in col.lower() or col.strip() == '']
    if unnamed_cols:
        issues['colonnes_sans_nom'] = unnamed_cols

    # Caractères spéciaux non standards
    special_char_cols = [col for col in df.columns if any(not c.isalnum() and c not in " _-()" for c in col)]
    if special_char_cols:
        issues['caractères_spéciaux_dans_colonnes'] = special_char_cols

    # Longueur anormale des noms de colonnes
    long_name_cols = [col for col in df.columns if len(col) > 80]
    if long_name_cols:
        issues['noms_colonnes_trop_longs'] = long_name_cols

    return df_name, issues

# Vérification des en-têtes
header_issues_dept = check_header_issues(df_departement, "Département")
header_issues_epci = check_header_issues(df_epci, "EPCI")

header_issues_dept, header_issues_epci


(('Département', {}), ('EPCI', {}))

#### Analyse: 

Aucun problème détecté au niveau des en-têtes pour les deux fichiers.
Plus précisément :

- Aucun nom de colonne ne contient d’espaces superflus.
- Toutes les colonnes ont un nom explicite (aucune colonne nommée "Unnamed" ou vide).
- Aucun caractère spécial problématique dans les noms de colonnes.
- Aucun nom de colonne n’est excessivement long.

Les en-têtes sont donc bien structurés et conformes pour une utilisation directe dans des analyses ou des traitements de données.

#### B- Vérification structurelle des valeurs

In [35]:
# Fonction de détection des problèmes structurels
def check_structural_issues(df, df_name):
    issues = {}

    # Vérifier les doublons de colonnes
    if df.columns.duplicated().any():
        issues['colonnes_dupliquées'] = df.columns[df.columns.duplicated()].tolist()

    # Vérifier les lignes dupliquées
    duplicated_rows = df[df.duplicated()]
    if not duplicated_rows.empty:
        issues['lignes_dupliquées'] = len(duplicated_rows)

    # Vérifier les valeurs manquantes
    missing_values = df.isnull().sum()
    cols_with_missing = missing_values[missing_values > 0]
    if not cols_with_missing.empty:
        issues['valeurs_manquantes'] = cols_with_missing.to_dict()

    # Vérifier l'unicité de certaines colonnes clés si pertinentes
    if 'Code département' in df.columns and 'Année' in df.columns:
        duplicate_keys = df.duplicated(subset=['Code département', 'Année']).sum()
        if duplicate_keys > 0:
            issues['doublons Code département + Année'] = duplicate_keys
    elif 'Code EPCI' in df.columns and 'Année' in df.columns:
        duplicate_keys = df.duplicated(subset=['Code EPCI', 'Année']).sum()
        if duplicate_keys > 0:
            issues['doublons Code EPCI + Année'] = duplicate_keys

    return df_name, issues

# Analyse des deux fichiers
dept_issues = check_structural_issues(df_departement, "Département")
epci_issues = check_structural_issues(df_epci, "EPCI")

dept_issues, epci_issues


(('Département', {}), ('EPCI', {'doublons Code département + Année': 13016}))

#### Analyse: 

Fichier Département

Aucun problème structurel détecté :
- Pas de colonnes dupliquées
- Pas de lignes dupliquées
- Pas de valeurs manquantes
- Clés (Code département + Année) uniques

Fichier EPCI

- Un problème détecté :
- Il existe 13 016 doublons sur les paires (Code département, Année). Cela signifie qu’un même département est présent plusieurs fois pour une même année dans ce fichier. Cela peut s’expliquer si un EPCI couvre plusieurs départements, mais cela mérite vérification selon l’usage des données.

#### C- Vérification des doublons du fichier EPCI

In [39]:
# Identification des doublons par EPCI, année et département
epci_duplicates = df_epci[df_epci.duplicated(subset=['Code département', 'Année'], keep=False)]

# Regrouper les doublons par Code département et Année pour identifier les cas récurrents
grouped_duplicates = (
    epci_duplicates
    .groupby(['Code département', 'Année'])[['Code EPCI', 'Libellé EPCI']]
    .agg(list)
    .reset_index()
)

# Affichage direct des premiers doublons avec print (limité à 5 lignes)
print(grouped_duplicates.head())

  Code département  Année                                          Code EPCI  \
0               01   2012  [240100750, 240100800, 200069193, 200070555, 2...   
1               01   2013  [200029999, 200042497, 200070118, 200071751, 2...   
2               01   2014  [200042935, 200040350, 200069193, 200071371, 2...   
3               01   2015  [200029999, 200070555, 200071751, 240100883, 2...   
4               01   2016  [240100800, 200042497, 200029999, 200042935, 2...   

                                        Libellé EPCI  
0  [CA du Pays de Gex, CC de Miribel et du Platea...  
1  [CC Rives de l'Ain - Pays du Cerdon, CC Dombes...  
2  [CA Haut - Bugey Agglomération, CC Bugey Sud, ...  
3  [CC Rives de l'Ain - Pays du Cerdon, CC de la ...  
4  [CC de Miribel et du Plateau, CC Dombes Saône ...  


#### Analyse:

Cela montre clairement que plusieurs intercommunalités couvrent un même département en une année donnée — une réalité administrative normale, mais qui nécessite une agrégation par département si l'on souhaite des analyses à cette échelle.

#### Prétraitements utiles

#### A- Agrégation du fichier EPCI

In [44]:
# Redéfinir les colonnes à agréger (numériques)
cols_to_aggregate = df_epci.select_dtypes(include='number').columns.tolist()

# Réaliser l'agrégation proprement
df_epci_grouped = (
    df_epci
    .groupby(['Code département', 'Année'], as_index=False)[cols_to_aggregate]
    .sum()
)

# Afficher un aperçu du résultat
df_epci_grouped.head()


Code département  Année   Code EPCI  Consommation agriculture (MWh)  \
0               01  28168  3001012703                        29896.12   
1               01  28182  3001012703                        32221.34   
2               01  28196  3001012703                        29361.67   
3               01  28210  3001012703                        48569.20   
4               01  28224  3001012703                        43842.96   

   Consommation industrie (MWh)  Consommation résidentiel (MWh)  \
0                    1528817.98                      2089436.57   
1                    1505125.66                      2164856.93   
2                    1494106.43                      1968833.02   
3                    2800960.66                      3572436.40   
4                    2823644.33                      3715740.29   

   Consommation tertiaire (MWh)  Consommation autre (MWh)  \
0                    2166381.37                  15481.50   
1                    1416303.83                  35278.53   
2                    1572445.35                  15282.56   
3                    2735830.68                  68274.65   
4                    2705991.56                  21222.09   

   Consommation totale (MWh)  Consommation électricité agriculture (MWh)  ...  \
0                 5830013.54                                    29896.12  ...   
1                 5153786.29                                    32221.34  ...   
2                 5080029.03                                    29361.67  ...   
3                 9226071.59                                    39963.20  ...   
4                 9310441.23                                    35489.96  ...   

   Consommation électricité résidentiel (MWh)  \
0                                  2089436.57   
1                                  2164856.93   
2                                  1968833.02   
3                                  2109810.40   
4                                  2138918.29   

   Consommation électricité tertiaire (MWh)  \
0                                2166381.37   
1                                1416303.83   
2                                1572445.35   
3                                2211993.68   
4                                2166635.56   

   Consommation électricité autre (MWh)  \
0                              15481.50   
1                              35278.53   
2                              15282.56   
3                              52576.65   
4                                869.09   

   Consommation électricité totale (MWh)  Consommation gaz agriculture (MWh)  \
0                             5830013.54                                 0.0   
1                             5153786.29                                 0.0   
2                             5080029.03                                 0.0   
3                             5911932.25                              8606.0   
4                             5861968.37                              8353.0   

   Consommation gaz industrie (MWh)  Consommation gaz résidentiel (MWh)  \
0                              0.00                                 0.0   
1                              0.00                                 0.0   
2                              0.00                                 0.0   
3                        1303372.34                           1462626.0   
4                        1303588.86                           1576822.0   

   Consommation gaz tertiaire (MWh)  Consommation gaz autre (MWh)  \
0                               0.0                           0.0   
1                               0.0                           0.0   
2                               0.0                           0.0   
3                          523837.0                       15698.0   
4                          539356.0                       20353.0   

   Consommation gaz totale (MWh)  
0                           0.00  
1                           0.00  
2                

#### Analyse: 

Le code agrège désormais correctement les données du fichier EPCI par Code département et Année, en additionnant toutes les colonnes numériques de consommation.
Chaque ligne du résultat représente maintenant la consommation totale d’un département pour une année donnée, ce qui est idéal pour :
- une analyse géographique à l’échelle départementale,
- une comparaison directe avec les données départementales officielles.

#### B- Exportation du fichier agrégé

In [48]:
# Export du fichier agrégé
aggregated_csv_path = r"conso-epci-par-departement-agregee.csv"
df_epci_grouped.to_csv(aggregated_csv_path, index=False)

aggregated_csv_path

'conso-epci-par-departement-agregee.csv'

#### 1.4 Nettoyage supplémentaire des données

#### A- Détection des erreurs par dataframe

In [52]:
# Fonction d'analyse des erreurs pour un DataFrame
def analyze_structural_issues(df, name):
    issues = {}

    # Valeurs manquantes
    missing = df.isnull().sum()
    missing_cols = missing[missing > 0]
    if not missing_cols.empty:
        issues['valeurs_manquantes'] = missing_cols.to_dict()

    # Doublons (sur toutes les colonnes)
    total_duplicates = df.duplicated().sum()
    if total_duplicates > 0:
        issues['doublons_complets'] = total_duplicates

    # Colonnes avec des types non standard
    non_numeric_cols = df.select_dtypes(exclude=['number', 'object']).columns.tolist()
    if non_numeric_cols:
        issues['colonnes_types_non_standard'] = non_numeric_cols

    # Vérification de la colonne Année
    if 'Année' in df.columns:
        invalid_years = df[~df['Année'].astype(str).str.match(r'^\d{4}$')]
        if not invalid_years.empty:
            issues['valeurs_invalides_Année'] = invalid_years['Année'].unique().tolist()

    # Vérification du Code département
    if 'Code département' in df.columns:
        invalid_dept_codes = df[~df['Code département'].astype(str).str.match(r'^\d{2,3}$')]
        if not invalid_dept_codes.empty:
            issues['valeurs_invalides_Code_département'] = invalid_dept_codes['Code département'].unique().tolist()

    return {name: issues}

# Appliquer l'analyse aux deux fichiers
issues_epci = analyze_structural_issues(df_epci_grouped, "EPCI Agrégé")
issues_departement = analyze_structural_issues(df_departement, "Département")

issues_epci | issues_departement


{'EPCI Agrégé': {'valeurs_invalides_Année': [28168,
   28182,
   28196,
   28210,
   28224,
   28238,
   28252,
   28266,
   28280,
   28294,
   28308,
   28322,
   38228,
   38247,
   38266,
   38285,
   38304,
   38323,
   38342,
   38361,
   38380,
   38399,
   38418,
   38437,
   20120,
   20130,
   20140,
   20150,
   20160,
   20170,
   20180,
   20190,
   20200,
   20210,
   20220,
   20230,
   10060,
   10065,
   10070,
   10075,
   10080,
   10085,
   10090,
   10095,
   10100,
   10105,
   10110,
   10115,
   14084,
   14091,
   14098,
   14105,
   14112,
   14119,
   14126,
   14133,
   14140,
   14147,
   14154,
   14161,
   32192,
   32208,
   32224,
   32240,
   32256,
   32272,
   32288,
   32304,
   32320,
   32336,
   32352,
   32368,
   16096,
   16104,
   16112,
   16120,
   16128,
   16136,
   16144,
   16152,
   16160,
   16168,
   16176,
   16184,
   26156,
   26169,
   26182,
   26195,
   26208,
   26221,
   26234,
   26247,
   26260,
   26273,
   26286,
   26299

#### Analyse: 
Voici les problèmes détectés dans les deux fichiers (EPCI Agrégé et Département) qui doivent être nettoyés pour garantir une exploitation statistique et géographique fiable :

Fichier EPCI Agrégé
 1. Année contient des valeurs invalides

- Des valeurs comme 28168, 10100, 30300 sont présentes.
- Ces valeurs ne sont pas des années valides (ex. 2020, 2021, etc.).
- Cela indique probablement un mauvais traitement ou une mauvaise colonne lors de l’agrégation.

Action : Identifier la bonne colonne source contenant l’année d’origine et la restaurer.

 2. Code département contient des formats incorrects

- Certains codes sont des chaînes multiples concaténées (ex : "01, 69", "2A", "94, 92, 91, 75, 95, 93").
- Cela résulte probablement de la concaténation de plusieurs départements lors de la fusion ou agrégation d’EPCI inter-départementaux.

Action :

- Soit filtrer uniquement les codes simples,
- Soit dupliquer les lignes pour répartir la consommation sur plusieurs départements si vous voulez la granularité exacte.

 Fichier Département

 1. Code département contient les codes "2A" et "2B"

- Bien que valides pour la Corse, ces codes peuvent poser problème si vos outils exigent un format strictement numérique (comme "20").

 Action :
 
- Soit conserver "2A" / "2B" si vos géodonnées les acceptent,
- Soit les recoder en "2A" → 20.1, "2B" → 20.2 ou "20" si besoin de normalisation.

#### B- Correction automatisée des erreurs

In [62]:
# --- Nettoyage du fichier EPCI Agrégé ---

# 1. Corriger la colonne Année : filtrer uniquement les valeurs valides à 4 chiffres entre 2000 et 2100
df_epci_cleaned = df_epci_grouped[df_epci_grouped['Année'].astype(str).str.match(r'^\d{4}$')]
df_epci_cleaned = df_epci_cleaned[df_epci_cleaned['Année'].astype(int).between(2000, 2100)]

# 2. Nettoyer Code département : ne garder que les codes simples (exclure ceux contenant des virgules)
df_epci_cleaned = df_epci_cleaned[~df_epci_cleaned['Code département'].astype(str).str.contains(',')]

# Reformatage : convertir en chaîne à deux caractères pour les codes valides
df_epci_cleaned['Code département'] = df_epci_cleaned['Code département'].astype(str).str.zfill(2)

# --- Nettoyage du fichier Département ---

# Reformatage : convertir en chaîne pour standardiser
df_departement_cleaned = df_departement.copy()
df_departement_cleaned['Code département'] = df_departement_cleaned['Code département'].astype(str)

# Normaliser les codes "2A" et "2B" si nécessaire (ici, on les laisse tels quels si compatibles)

# Aperçu des nettoyages effectués
df_epci_cleaned_preview = df_epci_cleaned.head()
df_departement_cleaned_preview = df_departement_cleaned.head()

df_epci_cleaned_preview, df_departement_cleaned_preview


(Empty DataFrame
 Columns: [Code département, Année, Code EPCI, Consommation agriculture (MWh), Consommation industrie (MWh), Consommation résidentiel (MWh), Consommation tertiaire (MWh), Consommation autre (MWh), Consommation totale (MWh), Consommation électricité agriculture (MWh), Consommation électricité industrie (MWh), Consommation électricité résidentiel (MWh), Consommation électricité tertiaire (MWh), Consommation électricité autre (MWh), Consommation électricité totale (MWh), Consommation gaz agriculture (MWh), Consommation gaz industrie (MWh), Consommation gaz résidentiel (MWh), Consommation gaz tertiaire (MWh), Consommation gaz autre (MWh), Consommation gaz totale (MWh)]
 Index: []
 
 [0 rows x 21 columns],
    Année Code département      Libellé département  Code région  \
 0   2012               02                    Aisne           32   
 1   2012               04  Alpes-de-Haute-Provence           93   
 2   2012               09                   Ariège           76   


#### Analayse: 

Le fichier départemental a été nettoyé avec succès et est prêt pour l’analyse.
En revanche, le fichier EPCI agrégé est vide après nettoyage, car aucune ligne ne contient une Année valide à 4 chiffres. Cela confirme que la colonne Année a été mal construite lors de l’agrégation.

Prochaine étape nécessaire :

- Il faut retourner au fichier d’origine (df_epci) pour :
- Repartir de la bonne colonne Année (celle initialement bien formatée),
- Refaire l’agrégation proprement, en préservant correctement Année comme un entier à 4 chiffres.

#### C- Ré-agrégation des données EPCI depuis le fichier source

In [66]:
# Refaire l’agrégation depuis le fichier source avec une vraie colonne Année
# S'assurer que Année est bien un entier à 4 chiffres
df_epci['Année'] = df_epci['Année'].astype(str).str.extract(r'(\d{4})').astype(int)

# Sélection des colonnes numériques pour l’agrégation
cols_to_aggregate = df_epci.select_dtypes(include='number').columns.tolist()
cols_to_aggregate.remove('Année')  # on exclut Année de l'agrégation

# Agrégation par Code département et Année
df_epci_grouped_fixed = (
    df_epci
    .groupby(['Code département', 'Année'], as_index=False)[cols_to_aggregate]
    .sum()
)

# Aperçu du résultat corrigé
df_epci_grouped_fixed.head()


Code département  Année   Code EPCI  Consommation agriculture (MWh)  \
0               01   2012  3001012703                        29896.12   
1               01   2013  3001012703                        32221.34   
2               01   2014  3001012703                        29361.67   
3               01   2015  3001012703                        48569.20   
4               01   2016  3001012703                        43842.96   

   Consommation industrie (MWh)  Consommation résidentiel (MWh)  \
0                    1528817.98                      2089436.57   
1                    1505125.66                      2164856.93   
2                    1494106.43                      1968833.02   
3                    2800960.66                      3572436.40   
4                    2823644.33                      3715740.29   

   Consommation tertiaire (MWh)  Consommation autre (MWh)  \
0                    2166381.37                  15481.50   
1                    1416303.83                  35278.53   
2                    1572445.35                  15282.56   
3                    2735830.68                  68274.65   
4                    2705991.56                  21222.09   

   Consommation totale (MWh)  Consommation électricité agriculture (MWh)  ...  \
0                 5830013.54                                    29896.12  ...   
1                 5153786.29                                    32221.34  ...   
2                 5080029.03                                    29361.67  ...   
3                 9226071.59                                    39963.20  ...   
4                 9310441.23                                    35489.96  ...   

   Consommation électricité résidentiel (MWh)  \
0                                  2089436.57   
1                                  2164856.93   
2                                  1968833.02   
3                                  2109810.40   
4                                  2138918.29   

   Consommation électricité tertiaire (MWh)  \
0                                2166381.37   
1                                1416303.83   
2                                1572445.35   
3                                2211993.68   
4                                2166635.56   

   Consommation électricité autre (MWh)  \
0                              15481.50   
1                              35278.53   
2                              15282.56   
3                              52576.65   
4                                869.09   

   Consommation électricité totale (MWh)  Consommation gaz agriculture (MWh)  \
0                             5830013.54                                 0.0   
1                             5153786.29                                 0.0   
2                             5080029.03                                 0.0   
3                             5911932.25                              8606.0   
4                             5861968.37                              8353.0   

   Consommation gaz industrie (MWh)  Consommation gaz résidentiel (MWh)  \
0                              0.00                                 0.0   
1                              0.00                                 0.0   
2                              0.00                                 0.0   
3                        1303372.34                           1462626.0   
4                        1303588.86                           1576822.0   

   Consommation gaz tertiaire (MWh)  Consommation gaz autre (MWh)  \
0                               0.0                           0.0   
1                               0.0                           0.0   
2                               0.0                           0.0   
3                          523837.0                       15698.0   
4                          539356.0                       20353.0   

   Consommation gaz totale (MWh)  
0                           0.00  
1                           0.00  
2                

#### Analyse: 

L'agrégation a été refaite correctement à partir de la colonne Année d'origine. Cette fois :
La colonne Année contient bien des valeurs valides à 4 chiffres (ex. 2012, 2013, etc.).
Les données sont agrégées par département et année, prêtes pour analyse.
Le fichier est structuralement propre, sauf potentiellement pour des Code département composites que nous pouvons corriger maintenant.

#### D- Nettoyage de la colonne code département

In [70]:
# Nettoyage final de la colonne Code département

# Supprimer les lignes contenant des codes multiples (avec virgule)
df_epci_grouped_cleaned = df_epci_grouped_fixed[~df_epci_grouped_fixed['Code département'].astype(str).str.contains(',')]

# Convertir les codes en chaîne et les formater sur 2 ou 3 caractères si nécessaire
df_epci_grouped_cleaned['Code département'] = df_epci_grouped_cleaned['Code département'].astype(str).str.zfill(2)

# Export du fichier final nettoyé
final_export_path = r"conso-epci-par-departement-agregee-nettoyee.csv"
df_epci_grouped_cleaned.to_csv(final_export_path, index=False)

final_export_path

C:\Users\baube\AppData\Local\Temp\ipykernel_15324\3705949100.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_epci_grouped_cleaned['Code département'] = df_epci_grouped_cleaned['Code département'].astype(str).str.zfill(2)


'conso-epci-par-departement-agregee-nettoyee.csv'

#### E- Fusion des dataframes et exportation du fichier

In [73]:
# Fusionner les deux fichiers sur les colonnes clés
# Garder toutes les lignes du fichier département (base de référence complète)
fusion_finale = pd.merge(
    df_departement_cleaned,
    df_epci_grouped_cleaned,
    on=['Code département', 'Année'],
    suffixes=('_département', '_epci'),
    how='left'
)

# Vérifier les colonnes fusionnées et aperçu
fusion_finale.columns, fusion_finale.head(3)

(Index(['Année', 'Code département', 'Libellé département', 'Code région',
        'Libellé région', 'Consommation agriculture (MWh)_département',
        'Consommation industrie (MWh)_département',
        'Consommation résidentiel (MWh)_département',
        'Consommation tertiaire (MWh)_département',
        'Consommation autre (MWh)_département',
        'Consommation totale (MWh)_département', 'Opérateur',
        'Consommation électricité agriculture (MWh)_département',
        'Consommation électricité industrie (MWh)_département',
        'Consommation électricité résidentiel (MWh)_département',
        'Consommation électricité tertiaire (MWh)_département',
        'Consommation électricité autre (MWh)_département',
        'Consommation électricité totale (MWh)_département',
        'Opérateur électricité',
        'Consommation gaz agriculture (MWh)_département',
        'Consommation gaz industrie (MWh)_département',
        'Consommation gaz résidentiel (MWh)_département',

#### Analayse:

La fusion est terminée. Le résultat combine les deux sources en un seul tableau :

- Colonnes du fichier départemental : identifiées avec le suffixe _département
- Colonnes du fichier EPCI agrégé : identifiées avec le suffixe _epci
- Les colonnes partagées (ex. consommation totale) sont alignées pour comparaison directe.

Ce fichier est maintenant prêt pour :
- Comparaison fine des données sources (EPCI vs Département)
- Visualisation cartographique des écarts
- Analyses statistiques enrichies avec les deux niveaux d’agrégation

In [76]:
# Exporter le fichier fusionné
fusion_export_path = r"fusion-departement-epci.csv"
fusion_finale.to_csv(fusion_export_path, index=False)

fusion_export_path

'fusion-departement-epci.csv'

#### 1.5 Nettoyage final du fichier fusionné

In [79]:
# Copie de travail pour finalisation
fusion_finale_cleaned = fusion_finale.copy()

# 1. Remplir les valeurs manquantes dans les colonnes _epci avec 0 (logique : "aucune donnée disponible")
epci_cols = [col for col in fusion_finale_cleaned.columns if col.endswith('_epci')]
fusion_finale_cleaned[epci_cols] = fusion_finale_cleaned[epci_cols].fillna(0)

# 2. Forcer les colonnes numériques à un type float
for col in epci_cols:
    fusion_finale_cleaned[col] = pd.to_numeric(fusion_finale_cleaned[col], errors='coerce')

# 3. Ajouter une colonne d’écart relatif (%) si données disponibles
fusion_finale_cleaned['Écart relatif (%)'] = np.where(
    (fusion_finale_cleaned['Consommation totale (MWh)_département'] > 0),
    100 * abs(
        fusion_finale_cleaned['Consommation totale (MWh)_département'] -
        fusion_finale_cleaned['Consommation totale (MWh)_epci']
    ) / fusion_finale_cleaned['Consommation totale (MWh)_département'],
    np.nan
)

# Export final du fichier 100% prêt
fusion_ready_path = r"fusion-departement-epci-clean.csv"
fusion_finale_cleaned.to_csv(fusion_ready_path, index=False)

fusion_ready_path

'fusion-departement-epci-clean.csv'

Le fichier fusionné est désormais 100% prêt pour des analyses statistiques et géographiques :

Améliorations apportées :

- Remplissage des valeurs manquantes _epci avec 0
- Conversion des colonnes _epci en nombres
- Ajout d’une colonne Écart relatif (%) pour comparer les consommations totales

#### 1.6 Création de nouvelles colonnes et présentation des données du fichier final

#### A- Création de nouvelles colonnes

In [89]:
df= pd.read_csv(r"fusion-departement-epci-clean.csv")
df.head()

Année Code département      Libellé département  Code région  \
0   2012               02                    Aisne           32   
1   2012               04  Alpes-de-Haute-Provence           93   
2   2012               09                   Ariège           76   
3   2012               10                     Aube           44   
4   2012               11                     Aude           76   

               Libellé région  Consommation agriculture (MWh)_département  \
0             Hauts-de-France                                    22131.31   
1  Provence-Alpes-Côte d'Azur                                    13914.73   
2                   Occitanie                                     3698.88   
3                   Grand Est                                    32844.98   
4                   Occitanie                                    35342.96   

   Consommation industrie (MWh)_département  \
0                                1149343.94   
1                                 211085.84   
2                                 255085.68   
3                                3075179.26   
4                                 845672.84   

   Consommation résidentiel (MWh)_département  \
0                                  1571035.97   
1                                   720302.92   
2                                   536606.68   
3                                  2318302.29   
4                                  2128257.24   

   Consommation tertiaire (MWh)_département  \
0                                 637364.89   
1                                 287308.03   
2                                 153944.85   
3                                 788109.56   
4                                 725948.72   

   Consommation autre (MWh)_département  ...  \
0                               3194.05  ...   
1                                658.96  ...   
2                               5001.57  ...   
3                             100068.08  ...   
4                               9391.74  ...   

   Consommation électricité tertiaire (MWh)_epci  \
0                                      637364.89   
1                                      136387.74   
2                                      153944.85   
3                                      389415.56   
4                                      480067.71   

  Consommation électricité autre (MWh)_epci  \
0                                   3194.05   
1                                    372.18   
2                                   5001.57   
3                                   1755.08   
4                                   2088.74   

   Consommation électricité totale (MWh)_epci  \
0                                  3382762.60   
1                                   663059.59   
2                                   954337.66   
3                                  2105299.68   
4                                  2206115.81   

   Consommation gaz agriculture (MWh)_epci  \
0                                     0.00   
1                                     0.00   
2                                     0.00   
3                                   266.00   
4                                 20041.92   

   Consommation gaz industrie (MWh)_epci  \
0                                   0.00   
1                                   0.00   
2                                   0.00   
3                             2423717.49   
4                              566826.10   

   Consommation gaz résidentiel (MWh)_epci  \
0                                      0.0   
1                                      0.0   
2                                      0.0   
3                                1288214.0   
4                                 667968.0   

   Consommation gaz tertiaire (MWh)_epci  Consommation gaz autre (MWh)_epci  \
0                                    0.0                                0.0   
1                                    0.0                                0.0   
2                                    0.0

#### B- Présentation des données

In [209]:
# Affichage des informations générales sur le DataFrame enrichi
fusion_info = fusion_finale_cleaned.info()
fusion_info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1209 entries, 0 to 1208
Data columns (total 52 columns):
 #   Column                                                  Non-Null Count  Dtype  
---  ------                                                  --------------  -----  
 0   Année                                                   1209 non-null   int64  
 1   Code département                                        1209 non-null   object 
 2   Libellé département                                     1209 non-null   object 
 3   Code région                                             1209 non-null   int64  
 4   Libellé région                                          1209 non-null   object 
 5   Consommation agriculture (MWh)_département              1209 non-null   float64
 6   Consommation industrie (MWh)_département                1209 non-null   float64
 7   Consommation résidentiel (MWh)_département              1209 non-null   float64
 8   Consommation tertiaire (MWh)_départeme

#### Analyse:

Voici les informations générales sur le fichier fusionné et enrichi :

 Structure du fichier

- Nombre de lignes : 1 209 (combinations département + année)
- Nombre de colonnes : 52
- Taille mémoire estimée : ~500 KB

 Types de données

- 42 colonnes numériques (float64)
- 8 colonnes textuelles (object)
- 2 colonnes entières (int64)

 Colonnes clés ajoutées pour l’analyse énergétique
 
- Consommation totale (MWh) : consommation départementale globale
- Consommation pilotable (MWh) : industrie + tertiaire
- Part pilotable (%) : part modulable
- Indice de flexibilité : indicateur synthétique

Le fichier est prêt pour tout type d’analyse, de statistique avancée ou de visualisation.

#### IX- Exportation du fichier

In [115]:
# Exporter le fichier enrichi avec les nouvelles colonnes
enriched_export_path = r"fusion-departement-epci-enrichi.csv"
fusion_finale_cleaned.to_csv(enriched_export_path, index=False)

enriched_export_path

'fusion-departement-epci-enrichi.csv'

## Deuxième partie: Analyse Exploratoire (EDA)

#### Objectif: créer un tableau de bord interactif, lisible et prêt à être présenté (Power BI).

#### 1- Statistiques descriptives

In [224]:
# Sélection des colonnes numériques pour statistiques descriptives
stats = fusion_finale_cleaned[[
    'Consommation totale (MWh)',
    'Consommation pilotable (MWh)',
    'Part pilotable (%)',
    'Indice de flexibilité'
]].describe()

stats

Consommation totale (MWh)  Consommation pilotable (MWh)  \
count               1.209000e+03                  1.209000e+03   
mean                8.793785e+06                  5.276767e+06   
std                 9.275746e+06                  6.897615e+06   
min                 2.736258e+05                  0.000000e+00   
25%                 2.983693e+06                  1.506905e+06   
50%                 6.359362e+06                  3.428879e+06   
75%                 1.058797e+07                  6.143549e+06   
max                 6.372949e+07                  4.636627e+07   

       Part pilotable (%)  Indice de flexibilité  
count         1209.000000            1209.000000  
mean            53.283687               8.279945  
std             13.991748              10.823271  
min              0.000000               0.000000  
25%             46.393051               2.364533  
50%             53.699803               5.380364  
75%             61.462935               9.640042  
max             87.183367              72.754807

#### Analayse: 

Voici les statistiques descriptives du fichier enrichi :

 Consommation totale (MWh) :
- Moyenne : 8,79 millions
- Médiane : 6,36 millions
- Min : 273 625
- Max : 63,7 millions

 Consommation pilotable (MWh) (industrie + tertiaire) :
- Moyenne : 5,28 millions
- Min : 0 (certaines zones sans industrie ou tertiaire)
- Max : 46,3 millions

 Part pilotable (%) :
- Moyenne : 53,3 %
- Min : 0 %
- Max : 87,2 %

Remarque : certains départements sont fortement pilotables, d’autres très peu.

 Indice de flexibilité (pondéré par conso totale) :
- Moyenne : 8,28
- Médiane : 5,38
- Max : 72,75

Cet indice met en valeur les départements à fort potentiel de modulation énergétique.